File: [4 kyu] Snail/main.cpp
#include <vector>

std::vector<int> snail(const std::vector<std::vector<int>> &snail_map) {
    std::vector<int> result;
    size_t row = snail_map.size();
    if (row == 0) {
        return result;
    }

    size_t col = snail_map[0].size();
    if (col == 0) {
        return result;
    }

    size_t total = row * col;
    int left = 0;
    int right = col - 1;
    int top = 0;
    int bottom = row - 1;

    while (result.size() < total) {
        for (int i = left; i <= right; i++) {
            result.push_back(snail_map[top][i]);
        }
        top++;

        for (int i = top; i <= bottom; i++) {
            result.push_back(snail_map[i][right]);
        }
        right--;

        for (int i = right; i >= left; i--) {
            result.push_back(snail_map[bottom][i]);
        }
        bottom--;

        for (int i = bottom; i >= top; i--) {
            result.push_back(snail_map[i][left]);
        }
        left++;
    }

    return result;
}

File: [4 kyu] Snail/main.py
def snail(snail_map):
    result = []
    row = len(snail_map)
    if row == 0:
        return result

    col = len(snail_map[0])
    if col == 0:
        return result

    total = row * col
    left, right = 0, col - 1
    top, bottom = 0, row - 1

    while len(result) < total:
        for i in range(left, right + 1):
            result.append(snail_map[top][i])
        top += 1

        for i in range(top, bottom + 1):
            result.append(snail_map[i][right])
        right -= 1

        for i in range(right, left - 1, -1):
            result.append(snail_map[bottom][i])
        bottom -= 1

        for i in range(bottom, top - 1, -1):
            result.append(snail_map[i][left])
        left += 1

    return result


def snail(snail_map):
    result = []

    while snail_map:
        result.extend(snail_map.pop(0))

        for row in snail_map:
            result.append(row.pop())

        if snail_map:
            result.extend(snail_map.pop()[::-1])

        for row in snail_map[::-1]:
            result.append(row.pop(0))

    return result
